### PR TITLE
Updates for Julia 1.0

### PR DESCRIPTION
**Pull request by ararslan**
Now `info` and `warn` are only imported and extended from Base when they're defined in Base. Travis and AppVeyor have been updated to handle versions 0.6 through 1.0.

Supersedes #100.

### CODE REVIEW

**Comment by ararslan** [comment]
Wow, the userimage tests segfault on 0.7 and 1.0. :grimacing:

**Comment by ararslan** [comment]
Since we're using TestSetExtensions in the tests here, we might need https://github.com/ssfrr/TestSetExtensions.jl/pull/7 before this will work properly. :confused:

File: .travis.yml [hunk 1]
```diff
@@ -25,9 +31,20 @@ matrix:
...
+        using Coverage
+        Codecov.submit(Codecov.process_folder())
+      '
+  - |
+      julia -e '
+        VERSION >= v"0.7.0-DEV.3656" && using Pkg
+        VERSION >= v"0.7.0-DEV.5183" || cd(Pkg.dir("Memento"))
+        Pkg.add("Documenter")
+        include(joinpath("docs", "make.jl"))
+      '
```
Comment (omus) [inline .travis.yml]: I sometimes combine these lines so I don't require the double VERSION checks

File: .travis/test.sh [hunk 2]
```diff
@@ -2,20 +2,63 @@
...
+    julia --color=yes --check-bounds=yes -e "$TESTCMD"
 elif [[ "$TEST_TYPE" == "userimage" ]]; then
     # Create a user image which includes Memento to make sure that _loggers is assigned at compile-time
     # See: https://github.com/invenia/Memento.jl/pull/21
-    julia -e '
-        LIB_PATH = abspath(Base.JULIA_HOME, Base.LIBDIR)
+    julia --color=yes -e '
+        if VERSION >= v"0.7.0-DEV.3382"
+            using Libdl
+        end
```
Comment (omus) [inline .travis/test.sh]: I'd make this a one-liner

File: src/records.jl [hunk 3]
```diff
@@ -83,17 +83,31 @@ Base.getindex(rec::Record, attr::Symbol) = getfield(rec, attr)
 Base.haskey(rec::T, attr::Symbol) where {T <: Record} = hasfield(T, attr)
 Base.keys(rec::T) where {T <: Record} = (fieldname(T, i) for i in 1:fieldcount(T))
 
-Base.start(rec::Record) = 0
-Base.done(rec::T, state) where {T <: Record} = state >= fieldcount(T)
+if isdefined(Base, :iterate)
+    function Base.iterate(rec::T, state=0) where T <: Record
+        state >= fieldcount(T) && return nothing
+        new_state = state + 1
```
Comment (omus) [inline src/records.jl]: Just reuse `state`?
```julia
state += 1
```

File: src/records.jl [hunk 4]
```diff
@@ -83,17 +83,31 @@ Base.getindex(rec::Record, attr::Symbol) = getfield(rec, attr)
...
 
-function Base.next(rec::T, state::Int) where T <: Record
-    new_state = state + 1
-    return (fieldname(T, new_state) => getfield(rec, new_state), new_state)
-end
+    function Base.iterate(rec::T, state=0) where T <: AttributeRecord
+        state >= fieldcount(T) && return nothing
+        new_state = state + 1
+        return (fieldname(T, new_state) => get(getfield(rec, new_state)), new_state)
+    end
```
Comment (omus) [inline src/records.jl]: These two iterate methods have the same body. Combine with `T <: Union{Record, AttributeRecord}`

File: test/handlers.jl [hunk 5]
```diff
@@ -106,7 +106,7 @@
 
         @testset "Sample Usage w/ File" begin
             filename = tempname()
-            info("Path to log file: $filename")
+            Compat.@info("Path to log file: $filename")
```
Comment (omus) [inline test/handlers.jl]: Personal preference but I prefer doing `using Compat: @info` so I don't have to have `Compat.*` peppered throughout the code. Will make dropping 0.6 support later easier.

File: .travis.yml (hunk 1 above)
Comment (ararslan) [reply .travis.yml]: I personally prefer having coverage submission and documentation deployment as separate steps, even if it means duplicating some logic until we drop 0.6.

File: src/records.jl (hunk 4 above)
Comment (omus) [reply src/records.jl]: Gross

File: test/io.jl [hunk 6]
```diff
@@ -9,7 +9,7 @@
             "fubar" => 50
         )
         roller_prefix = tempname()
-        info("Path to roller_prefix: $roller_prefix")
+        Compat.@info("Path to roller_prefix: $roller_prefix")
```
Comment (iamed2) [inline test/io.jl]: Doesn't have to be `Compat.` anymore since you imported it in test/runtests.jl right?

File: test/runtests.jl [hunk 7]
```diff
@@ -74,11 +75,15 @@ end
 
 _props(cr::ConstRecord) = (:level => cr[:level], :msg => cr[:msg])
 
-Base.start(cr::ConstRecord) = start(_props(cr))
-Base.next(cr::ConstRecord, state) = next(_props(cr), state)
-Base.done(cr::ConstRecord, state) = done(_props(cr), state)
+if isdefined(Base, :iterate)
+    Base.iterate(cr::ConstRecord, state=1) = iterate(_props(cr), state)
```
Comment (iamed2) [inline test/runtests.jl]: You can just do `state...` and pass it right through

File: test/io.jl (hunk 6 above)
Comment (ararslan) [reply test/io.jl]: Oh whoops, I missed that one.

File: test/runtests.jl (hunk 7 above)
Comment (ararslan) [reply test/runtests.jl]: > and pass it right through

How do you mean?

**Comment by ararslan** [comment]
> The userimg tests segfaulting might actually indicate issues with the library. The root cause should be identified and documented.

It's unrelated to Memento:

```
$ cat thingo.jl 
VERSION >= v"0.7.0-DEV.3382" && using Libdl
BINDIR = VERSION < v"0.7.0-DEV.3073" ? Base.JULIA_HOME : Sys.BINDIR
LIB_PATH = abspath(BINDIR, Base.LIBDIR)
sysimg_lib = joinpath(LIB_PATH, "julia", "sys.$(Libdl.dlext)")
userimg_o = "userimg.o"
userimg_lib = "userimg.$(Libdl.dlext)"
run(`$(Base.julia_cmd()) --output-o $userimg_o --sysimage $sysimg_lib --startup-file=no -e "println(1)"`)
$ julia-0.6 thingo.jl
1
$ ls
thingo.jl  userimg.o
$ julia-0.7 thingo.jl

signal (11): Segmentation fault
in expression starting at no file:0
uv_write2 at /buildworker/worker/package_linux64/build/deps/srccache/libuv-ed3700c849289ed01fe04273a7bf865340b2bd7e/src/unix/stream.c:1432
uv_write at /buildworker/worker/package_linux64/build/deps/srccache/libuv-ed3700c849289ed01fe04273a7bf865340b2bd7e/src/unix/stream.c:1524
jl_uv_write at /buildworker/worker/package_linux64/build/src/jl_uv.c:454
uv_write_async at ./stream.jl:810
uv_write at ./stream.jl:778
unsafe_write at ./stream.jl:834
macro expansion at ./gcutils.jl:87 [inlined]
write at ./strings/io.jl:152 [inlined]
show at ./show.jl:568 [inlined]
print at ./strings/io.jl:31
print at ./strings/io.jl:42
println at ./strings/io.jl:69
unknown function (ip: 0x7f77483c2507)
jl_fptr_trampoline at /buildworker/worker/package_linux64/build/src/gf.c:1829
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2182
println at ./coreio.jl:4
unknown function (ip: 0x7f77483c2453)
jl_fptr_trampoline at /buildworker/worker/package_linux64/build/src/gf.c:1829
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2182
do_call at /buildworker/worker/package_linux64/build/src/interpreter.c:324
eval_value at /buildworker/worker/package_linux64/build/src/interpreter.c:428
eval_stmt_value at /buildworker/worker/package_linux64/build/src/interpreter.c:363 [inlined]
eval_body at /buildworker/worker/package_linux64/build/src/interpreter.c:686
jl_interpret_toplevel_thunk_callback at /buildworker/worker/package_linux64/build/src/interpreter.c:799
unknown function (ip: 0xfffffffffffffffe)
unknown function (ip: 0x7f774b6c5b8f)
unknown function (ip: 0xffffffffffffffff)
jl_interpret_toplevel_thunk at /buildworker/worker/package_linux64/build/src/interpreter.c:808
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:831
jl_toplevel_eval_in at /buildworker/worker/package_linux64/build/src/builtins.c:633
eval at ./boot.jl:319
jl_fptr_trampoline at /buildworker/worker/package_linux64/build/src/gf.c:1829
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2182
macro expansion at ./logging.jl:317 [inlined]
exec_options at ./client.jl:229
_start at ./client.jl:432
jl_fptr_trampoline at /buildworker/worker/package_linux64/build/src/gf.c:1829
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2182
jl_apply at /buildworker/worker/package_linux64/build/ui/../src/julia.h:1538 [inlined]
true_main at /buildworker/worker/package_linux64/build/ui/repl.c:112
main at /buildworker/worker/package_linux64/build/ui/repl.c:233
__libc_start_main at /build/glibc-Cl5G7W/glibc-2.23/csu/../csu/libc-start.c:291
_start at /home/alex/Projects/past-julia/0.7/bin/julia (unknown line)
Allocations: 726978 (Pool: 726579; Big: 399); GC: 0
ERROR: LoadError: failed process: Process(`/home/alex/Projects/past-julia/0.7/bin/julia -Cnative -J/home/alex/Projects/past-julia/0.7/lib/julia/sys.so --compile=yes --depwarn=yes --output-o userimg.o --sysimage /home/alex/Projects/past-julia/0.7/lib/julia/sys.so --startup-file=no -e 'println(1)'`, ProcessSignaled(11)) [0]
Stacktrace:
 [1] error(::String, ::Base.Process, ::String, ::Int64, ::String) at ./error.jl:42
 [2] pipeline_error at ./process.jl:712 [inlined]
 [3] #run#509(::Bool, ::Function, ::Cmd) at ./process.jl:670
 [4] run(::Cmd) at ./process.jl:668
 [5] top-level scope at none:0
 [6] include at ./boot.jl:317 [inlined]
 [7] include_relative(::Module, ::String) at ./loading.jl:1038
 [8] include(::Module, ::String) at ./sysimg.jl:29
 [9] exec_options(::Base.JLOptions) at ./client.jl:239
 [10] _start() at ./client.jl:432
in expression starting at /home/alex/Projects/test/thingo.jl:7
```
Apparently as of https://github.com/JuliaLang/julia/pull/24410, the `--output-o` option is to be considered unstable and was hidden from the `--help` output. It seems that it's since bitrotted.

File: test/runtests.jl (hunk 7 above)
Comment (ararslan) [reply test/runtests.jl]: I see. Is that preferable?

**Comment by iamed2** [comment]
We'll need to change it to use `build_sysimg` I guess then

File: test/runtests.jl (hunk 7 above)
Comment (iamed2) [reply test/runtests.jl]: It doesn't assume iteration state, so yeah. Performance doesn't matter for this but I think there are splatting optimizations for small tuples.

**Comment by ararslan** [comment]
`build_sysimg` has also bitrotted. See https://github.com/JuliaLang/julia/pull/27629. The recommendation is to use PackageCompiler, but that too does not support 0.7 or newer.

**Comment by iamed2** [comment]
So the documented function doesn't work? https://docs.julialang.org/en/latest/devdocs/sysimg/#Building-the-Julia-system-image-1

That's...worrying

**Comment by ararslan** [comment]
I didn't realize we had it documented. I think once https://github.com/JuliaLang/julia/pull/21849 landed, rebuilding the sysimage became pretty uncommon, so no one paid much attention to the machinery in place that worked around the initial limitation.

File: .travis/test.sh [hunk 8]
```diff
@@ -2,20 +2,61 @@
 
 set -ev
 
+# Set up the build environment across Julia versions
+# This portion is the same as the default Travis test script
```
Comment (rofinn) [inline .travis/test.sh]: Maybe include a link to the default Travis test script?

File: .travis/test.sh [hunk 9]
```diff
@@ -2,20 +2,61 @@
...
+if [[ -a .git/shallow ]]; then
+    git fetch --unshallow
+fi
+if [[ -f Project.toml || -f JuliaProject.toml ]]; then
```
Comment (rofinn) [inline .travis/test.sh]: Is this condition entirely necessary while we don't have a Project.toml? Unrelated, but is there a reason there are two *Project.toml files now?

Comment (iamed2) [reply .travis/test.sh]: JuliaProject.toml is for when you have some other system that requires Project.toml. If it exists it'll be used instead.

File: .travis/test.sh (hunk 8 above)
Comment (omus) [reply .travis/test.sh]: https://github.com/travis-ci/travis-build/blob/master/lib/travis/build/script/julia.rb

File: .travis/test.sh (hunk 9 above)
Comment (ararslan) [reply .travis/test.sh]: No the condition isn't necessary but it seemed simpler to me to duplicate the Travis logic wholesale and not have to remember to update things here when we do add a Project.toml file.

**Comment by ararslan** [comment]
Well, I fixed the segfault but the userimage tests still don't work. ¯\\\_(ツ)\_/¯

**Comment by ararslan** [comment]
Maybe, but I'm not sure; in testing I've been using the stdlib Test rather than Memento, which should be easier for Julia to find, but it still errors saying it can't find it. Kristoffer was helping me diagnose the other day but we didn't come to a conclusion.

**Pull request merged**